lib/model/query/entities.js

```diff
@@ -62,55 +61,52 @@ select ins.*, def.id as "entityDefId", ds."acteeId" as "dsActeeId", ds."name" as
-                            .then((entity) => resolve(entity))
-                            .catch((err) => reject(Problem.user.entityViolation({ reason: 'Entity could not be created', err: err.problemDetails })))
-                          : reject(Problem.user.entityViolation({ reason: 'It is not possible to create an entity from this data.' })))))
-                  .catch((problem) => reject(problem)))))));
-  });
+const _processSubmissionDef = (submissionDefId) => async ({ Datasets, Entities, Submissions }) => {
+  const existingEntity = await Entities.getDefBySubmissionDefId(submissionDefId);
+  // If the submission has already been used to make an entity, don't try again
+  // and don't log it as an error.
+  if (existingEntity.isDefined())
+    return null;
+  const submissionDef = await Submissions.getDefById(submissionDefId).then((s) => s.get());
+  const fields = await Datasets.getFieldsByFormDefId(submissionDef.formDefId);
+  // No fields found for this formDefId means there is nothing entity-related to parse out
+  // of this submission. Even entity system properties like ID and dataset name would be here
+  // if the form def is about datasets and entities.
+  if (fields.length === 0)
+    return null;
+  const partial = await Entity.fromSubmissionXml(submissionDef.xml, fields);
+  const entity = await Entities.createNew(partial, submissionDef);
+  return entity;
+};
 
+const processSubmissionEvent = (event) => (container) =>
+  container.db.transaction((trxn) =>
+    container.with({ db: trxn }).Entities._processSubmissionDef(event.details.submissionDefId))
+    .then((entity) => {
+      if (entity != null) {
+        return container.Audits.log({ id: event.actorId }, 'entity.create', { acteeId: entity.aux.dataset.acteeId },
+          { entity: { uuid: entity.uuid, label: entity.label, dataset: entity.aux.dataset.name },
+            submissionId: event.details.submissionId,
+            submissionDefId: event.details.submissionDefId });
+      }
+    })
+    .catch((err) =>
+      // err could be any kind of problem, from an entity violation error, to a
+      // database constraint error, to some other kind of error within the processing code.
+      // We always surface the error message but only log the error if it is one of our
+      // known Problems, just in case there are weird error details we don't want to
+      // expose. In experimenting with breaking the code, it seems that non-Problem errors
+      // convert to empty objects through the JSON.stringify transformation within audit logging
+      // so this probably isn't even an issue.
+      // The JSON.stringify appears to strip out error.stack so that doesn't make it to the
+      // log details even for our Problems.
+      container.Audits.log({ id: event.actorId }, 'entity.create.error', null,
+        { submissionId: event.details.submissionId,
+          submissionDefId: event.details.submissionDefId,
+          errorMessage: err.message,
```

Comment:
It'd be nice to show this message in Frontend at some point in case the error isn't a Problem. I don't think we need to do that for v2022.3 though. Let me know if it'd be useful for me to file a Frontend issue about this!

Reply:
Yes! we should be able to update frontend to check for `errorMessage` instead. It will have slightly more detail (include the name of the problem concatenated with the reason) than just the currently displayed Problem reason.